File: cpp/questions/subarraySum.cpp
#include <unordered_map>
#include <vector>

using namespace std;

class Solution {
public:
    int subarraySum(vector<int>& nums, int k) {
        int n = nums.size();
        int ans = 0; 
        std::unordered_map<int, int> hm;
        int ps = 0;
        hm[ps]++;
        for (int i=0; i<n; ++i) {
            ps += nums[i];
            ans += hm[ps-k];
            hm[ps]++;
        }
        return ans;
    }
};

File: python/questions/subarraySum.py
from collections import defaultdict
from typing import List

class Solution:
    def subarraySum(self, nums: List[int], k: int) -> int:
        ans = 0
        dt = defaultdict(int)
        ps = 0
        dt[ps] += 1
        for n in nums:
            ps += n
            ans += dt[ps-k]
            dt[ps] += 1
        return ans
